csrc/multi_tensor_sgd_kernel.cu

```diff
@@ -160,6 +160,9 @@ void multi_tensor_sgd_cuda(
         TORCH_CHECK(tensor_lists[3][i].scalar_type() == at::ScalarType::Half,
                  "Additional output tensors should always be fp16.");
 
+  noop_flag=noop_flag.to(tensor_lists[0][0].device());
```

Comment:
Moving this to multi_tensor_apply.cuh would silently avoid the same crash for all other multi_tensor_apply clients.

However, maybe we want a crash.  Maybe we want to inform calling code noop_flag should be on the same device as the tensors, and give calling code the chance to put it there efficiently/permanently, rather than silently concealing the issue with `.to(tensor_lists[0][0].device());` every time the C++ backend is called.  If so, we should remove the above line and initialize
```python
self._dummy_overflow_buf = torch.cuda.IntTensor([0], device=self.param_groups[0]["params"][0].device)
```
In fact it's crazy ^ is not already present, I think we should add that regardless of C++ diffs.

Reply:
Sorry, forgot link to the Python line.  I think using `TORCH_CHECK(noop_flag.device() == tensor_lists[0][0].device(), "message of your choice")` here, together with changing [this line](https://github.com/NVIDIA/apex/blob/5b53121a2124d70dad0fd4f462d1392d06573ca9/apex/optimizers/fused_sgd.py#L101) in `FusedSGD.__init__` to use initialize it with the right device
```python
self._dummy_overflow_buf = torch.cuda.IntTensor([0], device=self.param_groups[0]["params"][0].device)
```
is the best approach if tests pass.